src/level/level.cpp

```diff
@@ -156,28 +181,80 @@ namespace level {
             Collidable *collision = getCollision(bulletNextPosition);
-
+            enums::CollisionType type = enums::CollisionType::NONE;
+            if (collision != nullptr) type = collision->getCollisionType();
             // TODO(simo): handle other types of collision
             // TODO(simo): memory leak quando collide con artifactti e powers
             // perché sono eliminati subito dopo la collisione
             // dovresti fare altre funzioni per gestire l'eliminazione di artefatti
             // e powers esternamente a questa classe
             // TODO(simo): getCollision dovrebbe essere const, e non fare altro
             // che ritornarti la collisione
-            if (collision != nullptr) {
+            if (type == enums::CollisionType::ENTITY) {
                 logger_.debug("bullet collision with %d", collision->getCollisionType());
+                if (bullets_[i]->handleEntityHit((Entity *) collision)) {
+                    deleteEnemy(collision);
+                }
                 bullets_[i]->clear(win);
                 delete bullets_[i];
                 bullets_.remove(i);
             } else {
-                bullets_[i]->move();
-                bullets_[i]->clearLast(win);
-                bullets_[i]->render(win);
+                if (type == enums::CollisionType::NONE) {
+                    bullets_[i]->move();
+                    bullets_[i]->clearLast(win);
+                    bullets_[i]->render(win);
+                } else {
+                    bullets_[i]->clear(win);
+                    delete bullets_[i];
+                    bullets_.remove(i);
+                }
                 i++;
             }
         }
     }
 
+    void Level::deleteEnemy(Collidable *collision) {
+        for (unsigned int i = 0; i < enemies_.size(); i++) {
+            if (collision == enemies_[i]) {
+                delete enemies_[i];
+                enemies_.remove(i);
+            }
+        }
+    }
+
+    void Level::enemiesAttack(WINDOW *win, manager::Level *levelManager) {
+        Position attackPos;
+        for (unsigned int i = 0; i < enemies_.size(); i++) {
+            for (int j = (enemies_[i]->getPosition().riga) - 1; j <= (enemies_[i]->getPosition().riga) + 1; j++) {
+                for (int k = (enemies_[i]->getPosition().colonna) - 1; k <= (enemies_[i]->getPosition().colonna) + 1; k++) {
+                    attackPos = Position(j, k);
+                    if ((levelManager->getPlayer()->getPosition()) == attackPos) {
+                        enemies_[i]->attack(levelManager->getPlayer());
+                        if (levelManager->getPlayer()->isDead()) {
+                            // TODO (gio?): implementare una finestra di gameover
+                        }
+                        enemies_[i]->clear(win);
+                        delete enemies_[i];
+                        enemies_.remove(i);
+                    }
+                }
+            }
+        }
```

Comment:
Se il raggio è 1 non ci possono essere elementi in mezzo, e ti basterebbe guardare la distanza con il player invece di utilizzare tutti i for annidati

Reply:
Poi mi crasha anche quando faccio la collisione :face_exhaling:

Reply:
Si rompe tutto perché con `enemies_.remove(i);` elimini l'enemies nel for più dentro, ma questo enemy è ancora utilizzato nel secondo for, quindi stai facendo una heap use after free, che mi fa crashare.

Reply:
@giospada secondo te il check se il player è morto bisogna farla nella view?? Sennò come fai a cambiare schermo??

Reply:
si sarebbe meglio farla arrivare fino alla view il messaggio

semplicemente se il player e morto puoi continuare l'esecuzione tanto è finito il gioco, appena torna alla view faccio il check

Reply:
Risolto il problema della collisione ed eliminato i 3 for utilizzando un modo più efficiente